File: Java/DecodeWaysII.java
public class DecodeWaysII {
    private static final long MOD = 1_000_000_007L;

    public int numDecodings(String s) {
        long prev2 = 1, prev1 = one(s.charAt(0));
        for (int i = 1; i < s.length(); i++) {
            long cur = (one(s.charAt(i)) * prev1 + two(s.charAt(i - 1), s.charAt(i)) * prev2) % MOD;
            prev2 = prev1;
            prev1 = cur;
        }
        return (int) (prev1 % MOD);
    }

    private static int one(char c) {
        if (c == '*') return 9;
        if (c == '0') return 0;
        return 1;
    }

    private static int two(char c1, char c2) {
        if (c1 == '1') return (c2 == '*') ? 9 : 1;
        if (c1 == '2') {
            if (c2 == '*') return 6;
            return (c2 >= '0' && c2 <= '6') ? 1 : 0;
        }
        if (c1 == '*') {
            if (c2 == '*') return 15;
            if (c2 >= '0' && c2 <= '6') return 2;
            return 1;
        }
        return 0;
    }

    public static void main(String[] args) {
        DecodeWaysII s = new DecodeWaysII();
        System.out.println(s.numDecodings("*"));  // 9
        System.out.println(s.numDecodings("1*")); // 18
        System.out.println(s.numDecodings("2*")); // 15
        System.out.println(s.numDecodings("0"));  // 0
        System.out.println(s.numDecodings("**")); // 96
        System.out.println(s.numDecodings("*1")); // 11
    }
}

File: Rust/src/bin/decode_ways_ii.rs
const MOD: i64 = 1_000_000_007;

fn one(c: u8) -> i64 {
    match c {
        b'*' => 9,
        b'0' => 0,
        _ => 1,
    }
}

fn two(c1: u8, c2: u8) -> i64 {
    match c1 {
        b'1' => if c2 == b'*' { 9 } else { 1 },
        b'2' => {
            if c2 == b'*' { 6 }
            else if (b'0'..=b'6').contains(&c2) { 1 }
            else { 0 }
        }
        b'*' => {
            if c2 == b'*' { 15 }
            else if (b'0'..=b'6').contains(&c2) { 2 }
            else { 1 }
        }
        _ => 0,
    }
}

fn num_decodings(s: String) -> i32 {
    let b = s.as_bytes();
    let (mut prev2, mut prev1) = (1i64, one(b[0]));
    for i in 1..b.len() {
        let cur = (one(b[i]) * prev1 + two(b[i - 1], b[i]) * prev2) % MOD;
        prev2 = prev1;
        prev1 = cur;
    }
    (prev1 % MOD) as i32
}

fn main() {
    println!("{}", num_decodings("*".to_string()));  // 9
    println!("{}", num_decodings("1*".to_string())); // 18
    println!("{}", num_decodings("2*".to_string())); // 15
    println!("{}", num_decodings("0".to_string()));  // 0
    println!("{}", num_decodings("**".to_string())); // 96
    println!("{}", num_decodings("*1".to_string())); // 11
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn examples() {
        assert_eq!(num_decodings("*".to_string()), 9);
        assert_eq!(num_decodings("1*".to_string()), 18);
        assert_eq!(num_decodings("2*".to_string()), 15);
        assert_eq!(num_decodings("0".to_string()), 0);
        assert_eq!(num_decodings("**".to_string()), 96);
        assert_eq!(num_decodings("*1".to_string()), 11);
    }
}
